bazaar/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+
+
+//! # Encointer Bazaar Module
+//!
+//! provides functionality for
+//! - creating new bazaar entries (shop and articles)
+//! - removing existing entries (shop and articles)
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+#[macro_use]
+extern crate approx;
+
+use frame_support::{
+    decl_event, decl_module, decl_storage, decl_error,
+    dispatch::DispatchResult,
+    ensure,
+    storage::{StorageDoubleMap, StorageMap},
+    traits::Get,
+    weights::{DispatchClass, Pays},
+    debug
+};
+use rstd::prelude::*;
+use sp_core::RuntimeDebug;
+use frame_system::ensure_signed;
+use sp_timestamp::OnTimestampSet;
+
+use rstd::{cmp::min, convert::TryInto};
+use codec::{Decode, Encode};
+use sp_runtime::traits::{Saturating, CheckedAdd, CheckedDiv, Zero, IdentifyAccount, Member, Verify, CheckedSub};
+use rstd::ops::Rem;
+
+use encointer_currencies::{CurrencyIdentifier, Location, Degree, LossyInto};
+use encointer_balances::BalanceType;
+use encointer_scheduler::{CeremonyIndexType, CeremonyPhaseType, OnCeremonyPhaseChange};
+
+pub trait Trait: frame_system::Trait 
+  //  + encointer_currencies::Trait 
+   // + encointer_balances::Trait 
+   // + encointer_scheduler::Trait
+{
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+    //type Public: IdentifyAccount<AccountId = Self::AccountId>;
+    //type Signature: Verify<Signer = Self::Public> + Member + Decode + Encode;
+}
+// Logger target
+const LOG: &str = "encointer";
+
+pub type ShopIdentifier = u64; //URL
+pub type ArticleIdentifier = u64; //URL
+/*
+#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, Default, RuntimeDebug)]
+pub struct ProofOfAttendance<Signature, AccountId> {
+    pub prover_public: AccountId,
+    pub ceremony_index: CeremonyIndexType,
+    pub currency_identifier: CurrencyIdentifier,
+    pub attendee_public: AccountId,
+    pub attendee_signature: Signature,
+}
+*/
+
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Bazaar {
+        // Maps the shop or article owner to the respective items 
+        pub ShopsOwned get(fn shops_owned): double_map hasher(blake2_128_concat) CurrencyIdentifier, hasher(blake2_128_concat) T::AccountId => Vec<ShopIdentifier>;
+        pub ArticlesOwned get(fn articles_owned): double_map hasher(blake2_128_concat) CurrencyIdentifier, hasher(blake2_128_concat) T::AccountId => Vec<ArticleIdentifier>; 
+        // Show item affiliation
+        pub ShopAffiliation get(fn shop_affiliation): double_map hasher(blake2_128_concat) CurrencyIdentifier, hasher(blake2_128_concat) ShopIdentifier => T::AccountId;
+        pub ArticleAffiliation get(fn article_affiliation): double_map hasher(blake2_128_concat) CurrencyIdentifier, hasher(blake2_128_concat) ArticleIdentifier => (T::AccountId, ShopIdentifier);
+        // The set of all shops and articles per currency (community)
+        pub ShopRegistry get(fn shop_registry): map hasher(blake2_128_concat) CurrencyIdentifier => Vec<ShopIdentifier>;
```

Comment:
This could become a very big Vec, but fir now that's fine I guess

Reply:
the bad thing about this is that the entire vec needs to be loaded and stored with each mutation. The alternative would be a map-set:
https://substrate.dev/recipes/map-set.html
but, again, lets leave it as it is for now